storage/remote/wal_watcher.go

```diff
@@ -0,0 +1,278 @@
+	if err != nil {
+		w.logger.Log("err", err)
+		return
+	}
+	w.readToEnd(segment)
+	// Send a message back to signal that we've read to the end of the current WAL segment and are starting to watch
+	// writes and new WAL segment files.
+	w.dirChan <- ""
+	var currentSegment = fmt.Sprintf("%08d", n)
+	defer segment.Close()
+	segmentRegex := regexp.MustCompile(`\d{8}`)
+	checkpointRegex := regexp.MustCompile(`checkpoint.\d{6}`)
+
+	err = watcher.Add(w.walDir)
+	if err != nil {
+		level.Error(w.logger).Log("err", err)
+		return
+	}
+
+	for {
+		select {
+		case <-w.quit:
+			return
+		// TODO: callum, handle maintaining the WAL pointer somehow across apply configs?
+		case event, ok := <-watcher.Events:
+			_, fileName := filepath.Split(event.Name)
+
+			if !ok {
+				return
+			}
+			if event.Op&fsnotify.Write == fsnotify.Write {
+				if fileName != currentSegment {
+					continue
+				}
+				w.decodeSegment(segment)
+			}
+			if event.Op&fsnotify.Create == fsnotify.Create {
+				if ok := segmentRegex.MatchString(fileName); ok {
+					// TODO: callum, ensure the create was for the segment n+1 and not n+2 or more
+					// Read the current segment one last time in case we receive events out of order. Not sure if this is necessary.
+					segment.Close()
+					n++
+
+					segment, err = wal.OpenReadSegment(wal.SegmentName(w.walDir, n))
+					if err != nil {
+						w.logger.Log("err", err)
+						return
+					}
+					currentSegment = fmt.Sprintf("%08d", n)
+				}
+				if ok := checkpointRegex.MatchString(fileName); ok {
+					w.logger.Log("msg", "checkpoint created", "checkpoint", event.Name)
+					// Head was truncated and WAL segments were checkpointed, so we should trim older series
+					for q := range w.storage.queues {
+						q.trimSeries()
+					}
+				}
+			}
+		case err, ok := <-watcher.Errors:
+			if !ok {
+				return
+			}
+			level.Error(w.logger).Log("err", err)
+		}
+	}
+}
+
+// Blocks  until the sample is sent to all remote write endpoints
+func (w *WALWatcher) forwardSamples(s []tsdb.RefSample) error {
+	// samplesCopy := s[:]
+	w.storage.queuesMtx.Lock()
+	tempQueues := make(map[*QueueManager]struct{})
+	for q := range w.storage.queues {
+		tempQueues[q] = w.storage.queues[q]
+	}
+	w.storage.queuesMtx.Unlock()
+	// ticker := time.NewTicker(1 * time.Second)
+	for _, sample := range s {
+		queuesCopy := tempQueues
+		for {
+			select {
+			case <-w.exitForwarding:
+				return fmt.Errorf("exit when attempting to forward a sample")
+			default:
+
+				for q := range queuesCopy {
+					ok := q.Append(sample)
+					if ok {
+						delete(queuesCopy, q)
+						continue
+					}
+				}
+				if len(queuesCopy) == 0 {
+					return nil
+				}
+
+			}
+		}
+	}
+	return nil
+}
+
+func (w *WALWatcher) decodeSegment(segment *wal.Segment) {
+	// todo: callum, is there an easy way to detect if r.Next() has returned the last possible record in a segment.
+	// For now just recover from the panic that is thrown if we call r.Next() when we shouldn't have
```

Comment:
re-reading something that EOF'd should just EOF; never panic in normal operations.

Reply:
tsdb pr required